### ISSUE TITLE
Cancel Listing

### ISSUE THREAD

**Issue by pamungkaski** [ProjectOpenSea/seaport#533]
Regarding Cancel
The code
https://github.com/ProjectOpenSea/seaport/blob/171f2cd7faf13b2bf0455851499f1981274977f7/contracts/lib/OrderValidator.sol#L358-L360

```
// Update the order status as not valid and cancelled.
orderStatus.isValidated = false;
orderStatus.isCancelled = true;
```

So cancelation is only marking this hash to be cancled and not active, which means if the user already lowered a listing several time, like a true degen, they will need to cancel every single listing they made while lowering. Is it possible to use a nonce canceling instead of hash canceling, so basically each lowered listing will have the same nonce, hence when once is cancelled on chain, every single one of them is also canceled.

I am thinking of doing that cancelation on the zone. but I am kinda curious why we don't use this nonce cancelation method instead of hash canceling cus it kinda save more gas. CMIIW

**Comment by njovy** [comment]
The zone of the order is an optional secondary account attached to the order with two additional privileges:
The zone may cancel orders where it is named as the zone by calling cancel. (Note that offerers can also cancel their own orders, either individually or for all orders signed with their current counter at once by calling incrementCounter).
"Restricted" orders (as specified by the order type) must either be executed by the zone or the offerer, or must be approved as indicated by a call to an isValidOrder or isValidOrderIncludingExtraData view function on the zone.

As the doc says there is a function ```incrementCounter``` that you can use to cancel all orders signed.

**Comment by pamungkaski** [comment]
> The zone of the order is an optional secondary account attached to the order with two additional privileges: The zone may cancel orders where it is named as the zone by calling cancel. (Note that offerers can also cancel their own orders, either individually or for all orders signed with their current counter at once by calling incrementCounter). "Restricted" orders (as specified by the order type) must either be executed by the zone or the offerer, or must be approved as indicated by a call to an isValidOrder or isValidOrderIncludingExtraData view function on the zone. https://docs.opensea.io/v2.0/reference/seaport-overview As the doc says there is a function `incrementCounter` that you can use to cancel all orders signed.

yup, but there is a case in which i only want to cancel order for a single NFT. the incrementCounter invalidate all order signed for that wallet, hence it cannot be used, I still need to cancel my order 1 by 1 on a single NFT.

Yeah, so my idea is to implement another way to cancel all order an NFT has, for now I am going to use zone to cancel that order probably.

**Comment by njovy** [comment]
I guess you misunderstand of a counter with a nonce. Unlike a nonce, Seaport uses a counter for cancelling multiple orders that the offerer created with one transaction. If you create a new order you can see Seaport didn't update a counter so it uses the same counter till it's increased. Each order use a salt for uniqueness. The counter isn't for a nonce so what you asked cannot be done.

**Comment by njovy** [comment]
You referred a counter with a nonce. Seaport doesn't use a counter as a nonce. They use a salt for a unique hash so cancel by a nonce is pretty much same as incrementCounter.

**Comment by pamungkaski** [comment]
> You referred a counter with a nonce. Seaport doesn't use a counter as a nonce. They use a salt for a unique hash so cancel by a nonce is pretty much same as incrementCounter.

yeah thats exactly my point, why didnt seaport use nonce cancelation cus it will safe gas to cancel all order on the same NFT, rather they do cancelation for each hash, which means I need to cancel each order 1 by 1 if i dont want to cancel other NFT' sorder.

As a NFT tradder myself, I often found myself lowering the price then needed to cancel all of the order that I created 1 by 1. I think canceling all order in 1 NFT is much more common to do compared canceling all active order i ever signed

**Comment by njovy** [comment]
As far as I know the purpose of Seaport is to minimize gas fee as possible as they can and they did decrease gas limit compared to Wyvern.
What you suggest will require an extra field, storage, or calculation to have the same nonce for each NFT. LooksRare uses this approach and it takes an extra map for keep track of what you suggested. People who don't require to call a cancellation will waste gas fee. Sending a cancellation transaction is only required when they want to list a higher price and Seaport already provide a way to cancel multiple orders with a minimum cost. 
I believe Seaport made the right design to minimize the gas cost but if you have a better option than the current version that can save the gas cost you should make a pull request! 
Cancelling one item costs more than incrementCounter so if a user really care about money(gas fee), the user would use incrementCounter even though relisting is required.

**Comment by pamungkaski** [comment]
> As far as I know the purpose of Seaport is to minimize gas fee as possible as they can and they did decrease gas limit compared to Wyvern. What you suggest will require an extra field, storage, or calculation to have the same nonce for each NFT. This will also affect users when they fulfill orders to lookup this nonce. LooksRare uses this approach (having an order nonce) and it takes an extra map for keep track of what you suggested. People who don't require to call a cancellation will waste gas fee. Sending a cancellation transaction is only required when they want to list a higher price and Seaport already provide a way to cancel multiple orders with a minimum cost. I believe Seaport made the right design to minimize the gas cost but if you have a better option than the current version that can save the gas cost you should make a pull request! Cancelling one item costs more than incrementCounter so if a user really cares about money(gas fee), the user would use incrementCounter even though relisting is required.

CMIIW but seaport save the whole hash to mark an order is canceled
https://github.com/ProjectOpenSea/seaport/blob/171f2cd7faf13b2bf0455851499f1981274977f7/contracts/lib/OrderValidator.sol#L29-L30

IMO related to storage, I don't think it is that different saving canceled in a hash compared to storing nonce. I think that in worst case scenario, storing it in hash take more storage compared to have a single nonce since you have to store every single order status if the canceled. 

Probably it is more of a way to match how wyvern works. Since wyverns, on OpenSea, when you want to lower the price, you will create a new order with lower price, this order will then appended to the current active order of the NFT. While if you use nonce, lowering price means replacing the current active order with new one by assigning the new nonce. Tho you can still have multiple order active, it just that when once is canceled other will be also canceled.
